OpenRA.Game/Map/MapGrid.cs

```diff
@@ -96,32 +136,44 @@ public MapGrid(MiniYaml yaml)
 			}
 
-			var makeCorners = Type == MapGridType.RectangularIsometric ?
-				(Func<int[], WVec[]>)IsometricCellCorners : RectangularCellCorners;
-			CellCorners = cellCornerHalfHeights.Select(makeCorners).ToArray();
-			TilesByDistance = CreateTilesByDistance();
-		}
-
-		static WVec[] IsometricCellCorners(int[] cornerHeight)
-		{
-			return new WVec[]
+			// Slope types are hardcoded following the convention from the TS and RA2 map format
+			Ramps = new[]
 			{
-				new WVec(-724, 0, 724 * cornerHeight[0]),
-				new WVec(0, -724, 724 * cornerHeight[1]),
-				new WVec(724, 0, 724 * cornerHeight[2]),
-				new WVec(0, 724, 724 * cornerHeight[3])
+				// Flat
+				new CellRamp(Type),
+
+				// Two adjacent corners raised by half a cell
+				new CellRamp(Type, bl: RampCornerHeight.Half, br: RampCornerHeight.Half),
+				new CellRamp(Type, tl: RampCornerHeight.Half, bl: RampCornerHeight.Half),
+				new CellRamp(Type, tl: RampCornerHeight.Half, tr: RampCornerHeight.Half),
+				new CellRamp(Type, tr: RampCornerHeight.Half, br: RampCornerHeight.Half),
+
+				// One corner raised by half a cell
+				new CellRamp(Type, bl: RampCornerHeight.Half, split: RampSplit.Y),
+				new CellRamp(Type, tl: RampCornerHeight.Half, split: RampSplit.X),
+				new CellRamp(Type, tr: RampCornerHeight.Half, split: RampSplit.Y),
+				new CellRamp(Type, br: RampCornerHeight.Half, split: RampSplit.X),
+
+				// Three corners raised by half a cell
+				new CellRamp(Type, tl: RampCornerHeight.Half, bl: RampCornerHeight.Half, br: RampCornerHeight.Half, split: RampSplit.Y),
+				new CellRamp(Type, tl: RampCornerHeight.Half, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half, split: RampSplit.X),
+				new CellRamp(Type, tl: RampCornerHeight.Half, tr: RampCornerHeight.Half, br: RampCornerHeight.Half, split: RampSplit.Y),
+				new CellRamp(Type, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half, br: RampCornerHeight.Half, split: RampSplit.X),
+
+				// Full tile sloped (mid corners raised by half cell, far corner by full cell)
+				new CellRamp(Type, tl: RampCornerHeight.Half, bl: RampCornerHeight.Full, br: RampCornerHeight.Half),
+				new CellRamp(Type, tl: RampCornerHeight.Full, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half),
+				new CellRamp(Type, tl: RampCornerHeight.Half, tr: RampCornerHeight.Full, br: RampCornerHeight.Half),
+				new CellRamp(Type, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half, br: RampCornerHeight.Full),
+
+				// Two opposite coners raised by half a cell
+				new CellRamp(Type, tl: RampCornerHeight.Half, br: RampCornerHeight.Half, split: RampSplit.X),
+				new CellRamp(Type, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half, split: RampSplit.Y),
+				new CellRamp(Type, tl: RampCornerHeight.Half, br: RampCornerHeight.Half, split: RampSplit.X),
+				new CellRamp(Type, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half, split: RampSplit.Y),
```

Comment:
The X and Y splits in the last two lines here should be reversed.

Reply:
It's rotational symmetry; following the same rotation pattern as the other blocks.
```c#
new CellRamp(Type, tl: RampCornerHeight.Half, br: RampCornerHeight.Half, split: RampSplit.X),
new CellRamp(Type, tr: RampCornerHeight.Half, bl: RampCornerHeight.Half, split: RampSplit.Y),
new CellRamp(Type, br: RampCornerHeight.Half, tl: RampCornerHeight.Half, split: RampSplit.X),
new CellRamp(Type, bl: RampCornerHeight.Half, tr: RampCornerHeight.Half, split: RampSplit.Y),
```

Reply:
It is actually supposed to be XXYY, so all of us were wrong :)